### PR TITLE
Implemented Add Text Feature

### LINKED ISSUE #746: Add 'Add Text' option
Add `Add Text` option in `Enhance Created PDFs` section.

There should be an option to `Select PDF file`, `Select Text File` and `Create PDF` button.
On clicking `Create PDF` button new PDF should be created by appending the text in a text file to selected PDF.

Maybe this can be useful : [https://www.javamadesoeasy.com/2016/06/how-to-modify-add-text-to-existing-pdf.html](https://www.javamadesoeasy.com/2016/06/how-to-modify-add-text-to-existing-pdf.html)

### PR DESCRIPTION
**Pull request by jddeep**
# Description

Implemented Add Text Feature.
Screenshot_20190502-220113
Screenshot_20190502-220122
Screenshot_20190502-220128
Screenshot_20190502-220145
Screenshot_20190502-220159

Fixes #746 

## Type of change
Just put an x in the [] which are valid.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.
- [x] `./gradlew assembleDebug assembleRelease`
- [x] `./gradlew checkstyle`

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by jddeep** [comment]
@vyankatesh24 Please review.

**Comment by jddeep** [comment]
@vyankatesh24 I could not understand the error for the build failure here?

File: app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+            getSnackbarwithAction(mActivity, R.string.snackbar_pdfCreated)
+                    .setAction(R.string.snackbar_viewAction, v -> mFileUtils.openFile(mPath))
+                    .show();
+        } catch (Exception e) {
+            e.printStackTrace();
+        } finally {
+            mMorphButtonUtility.initializeButton(mSelectPDF, mCreateTextPDF);
+            mMorphButtonUtility.initializeButton(mSelectText, mCreateTextPDF);
+            mTextPath = null;
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java]: You are doing this in `resetview` method also.

**Comment by jddeep** [comment]
@vyankatesh24 Please review. Can we keep the addition to favourites in a separate beginner issue for other new contributors?.

File: app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    }
+
+    public void resetView() {
+        mPdfpath = mTextPath = null;
+        mPdfUri = mTextUri = null;
+        mMorphButtonUtility.morphToGrey(mCreateTextPDF, mMorphButtonUtility.integer());
+        mCreateTextPDF.setEnabled(false);
+    }
+
+    private void addText(String fileName) {
```
Comment (Swati4star) [inline app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java]: add comments to functions

Comment (jddeep) [reply app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java]: sure

**Comment by jddeep** [comment]
> @jddeep
> 
> * Text is not getting added properly (only single line is getting added, try with the large file )
> * Text should be appended to previous content (it gets added at the start)
> * Added Text is getting formatted. (becoming bold).
> * Both file selection buttons gets `select a file` text after PDF is created. Change as per its function.
> * Bottom sheet is missing.

@vyankatesh24 Bottom sheet will be of no use here as we need to select two different kinds of files at the same time. Plus i used the resource you mentioned in the issue for adding the text. If you think this should be changed, then i need to do more research work. What do you want?

**Comment by vyankatesh24** [comment]
> > @jddeep
> > 
> > * Text is not getting added properly (only single line is getting added, try with the large file )
> > * Text should be appended to previous content (it gets added at the start)
> > * Added Text is getting formatted. (becoming bold).
> > * Both file selection buttons gets `select a file` text after PDF is created. Change as per its function.
> > * Bottom sheet is missing.
> 
> @vyankatesh24 Bottom sheet will be of no use here as we need to select two different kinds of files at the same time. Plus i used the resource you mentioned in the issue for adding the text. If you think this should be changed, then i need to do more research work. What do you want?

* Bottom sheet will be to select PDF files only.
* I did not get this `Plus i used the resource you mentioned in the issue for adding the text. If you think this should be changed, then i need to do more research work. What do you want?`

**Comment by jddeep** [comment]
@vyankatesh24 Please review. All changes were made. But two problems remain.
1. The text still appears as in a SINGLE LINE.
2. The up and down arrow buttons of the bottom sheet are not functional.
Can you please test the above and make any suggestions on how to fix them? @vyankatesh24

**Comment by vyankatesh24** [comment]
> 2. The up and down arrow buttons of the bottom sheet are not functional.
>    Can you please test the above and make any suggestions on how to fix them? @vyankatesh24

You haven't written code for that. 
Write the OnClick method for that .

File: app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+            // Add text in existing PDF
+            pdfContentByte.beginText();
+            pdfContentByte.setFontAndSize(BaseFont.createFont
+                            (BaseFont.TIMES_ROMAN, //Font name
+                                    BaseFont.CP1257, //Font encoding
+                                    BaseFont.EMBEDDED //Font embedded
+                            )
+                    , 12); // set font and size
+            pdfContentByte.setTextMatrix(10, 10); // set x and y co-ordinates
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java]: for single line problem try removing this.
Refer `itext` documentation, see `PdfContentByte` class.
Try if can use `Paragraph` class. Refer `TextToPDF` creation for this.

File: app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    public void resetView() {
+        mPdfpath = mTextPath = null;
+        mPdfUri = mTextUri = null;
+        mMorphButtonUtility.morphToGrey(mCreateTextPDF, mMorphButtonUtility.integer());
+        mCreateTextPDF.setEnabled(false);
+    }
+
+    private void addText(String fileName) {
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java]: Put adding text on the background thread and show animation as well.

File: app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+import static swati4star.createpdf.util.StringUtils.getDefaultStorageLocation;
+import static swati4star.createpdf.util.StringUtils.getSnackbarwithAction;
+import static swati4star.createpdf.util.StringUtils.showSnackbar;
+
+public class AddTextFragment extends Fragment implements MergeFilesAdapter.OnClickListener,
+        BottomSheetPopulate, OnBackPressedInterface {
+    private Activity mActivity;
+    private String mPdfpath;
+    private String mTextPath;
+    private Uri mPdfUri;
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java]: Are these variables needed `mPdfUri` and `mTextUri`?

File: app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+import static swati4star.createpdf.util.Constants.textExtension;
+import static swati4star.createpdf.util.DialogUtils.createOverwriteDialog;
+import static swati4star.createpdf.util.StringUtils.getDefaultStorageLocation;
+import static swati4star.createpdf.util.StringUtils.getSnackbarwithAction;
+import static swati4star.createpdf.util.StringUtils.showSnackbar;
+
+public class AddTextFragment extends Fragment implements MergeFilesAdapter.OnClickListener,
+        BottomSheetPopulate, OnBackPressedInterface {
+    private Activity mActivity;
+    private String mPdfpath;
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java]: make `p` to uppercase in `path`.

**Comment by vyankatesh24** [comment]
If Text files selected first, snackbar appears, `Error Occurred`.

**Comment by sidhuparas** [comment]
Any progress on this @jddeep?

**Comment by jddeep** [comment]
> Any progress on this @jddeep?

@sidhuparas I had communicated with @vyankatesh24 about the continuation of this. Please get in touch with him once. Thanks.

**Comment by vyankatesh24** [comment]
Yeah, @sidhuparas 
@jddeep has informed me that he will not be able to work further on it.

**Comment by jddeep** [comment]
@vyankatesh24 The one-line text error is fixed. Please get it merged. Thanks.

File: app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java [hunk 7]
```diff
@@ -13,6 +13,7 @@
 import android.support.v4.app.Fragment;
 import android.support.v7.preference.PreferenceManager;
 import android.support.v7.widget.RecyclerView;
+import android.util.Log;
```
Comment (vyankatesh24) [inline app/src/main/java/swati4star/createpdf/fragment/AddTextFragment.java]: Remove log statements

**Comment by jddeep** [comment]
@vyankatesh24 Changes made. 👍

**Comment by jddeep** [comment]
@vyankatesh24 Conflicts resolved too. Please merge. Thanks.

**Pull request merged**